src/monitoring/index.ts

```diff
@@ -0,0 +1,295 @@
+      auth: SYSTEM_GITHUB_TOKEN,
+      request: { retries: 5 },
+    })
+
+    return ResultAsync.fromPromise(
+      octokitWithRetry.request(
+        "GET /repos/opengovsg/isomer-redirection/contents/src/certbot-websites.csv"
+      ),
+      (error) =>
+        new MonitoringError(`Failed to fetch redirection domains: ${error}`)
+    )
+      .andThen((response) => {
+        const content = Buffer.from(response.data.content, "base64").toString(
+          "utf-8"
+        )
+        return ResultAsync.fromPromise(
+          new Promise<RedirectionDomain[]>((resolve, reject) => {
+            Papa.parse(content, {
+              header: true,
+              complete(results) {
+                // validate the csv
+                if (!results.data) {
+                  reject(new MonitoringError("Failed to parse csv"))
+                }
+                resolve(results.data as RedirectionDomain[])
+              },
+              error(error: unknown) {
+                reject(error)
+              },
+            })
+          }),
+          (error) => new MonitoringError(`Failed to parse csv: ${error}`)
+        )
+      })
+      .map((redirectionDomains) =>
+        redirectionDomains
+          .map((domain) => domain.source)
+          .map(
+            (domain) =>
+              ({
+                domain,
+                type: IsomerHostedDomainType.REDIRECTION,
+              } as IsomerHostedDomain)
+          )
+      )
+  }
+
+  /**
+   * This is in charge of fetching all the domains that are are under Isomer, inclusive
+   * of any subdomains and redirects.
+   */
+  getAllDomains() {
+    this.monitoringServiceLogger.info("Fetching all domains")
+    return ResultAsync.combine([
+      this.getAmplifyDeployments().mapErr(
+        (error) => new MonitoringError(error.message)
+      ),
+      this.getRedirectionDomains(),
+      this.getKeyCdnDomains(),
+    ]).andThen(([amplifyDeployments, redirectionDomains, keyCdnDomains]) => {
+      this.monitoringServiceLogger.info("Fetched all domains")
+      return okAsync(
+        [...amplifyDeployments, ...redirectionDomains, ...keyCdnDomains].sort(
+          (a, b) => {
+            const domainA = a.domain
+            const domainB = b.domain
+            if (
+              domainA.startsWith("www.") &&
+              domainA.slice(`www.`.length) === domainB
+            ) {
+              return 0
+            }
+            if (
+              domainB.startsWith("www.") &&
+              domainA === domainB.slice(`www.`.length)
+            ) {
+              return 0
+            }
+            if (domainA === domainB) return 0
+            return domainA > domainB ? 1 : -1
+          }
+        )
+      )
+    })
+  }
+
+  // todo: once /siteup logic is merged into dev, we can add that as to alert isomer team
+  generateReportCard(domains: IsomerHostedDomain[]) {
+    const reportCard: ReportCard[] = []
+
+    const domainResolvers = domains.map(({ domain, type }) => {
```

Comment:
actually this function appears to be doing domain resolution and reporting the resolved records. 

should we be splitting the method up so it's clearer?

Reply:
yeap, i think in my head, the flow is something like: 
```ts
generateReportcard = (domains) => { 
  generateARecords().andThen(generateQuadA).andThen(generateCaa).andThen(generateCname).orElse(someError)
}
```

but this would potentially be expensive to do. ok w/ just shifting the domains out: 

```ts
return generateDomains().andThen(() => {
        this.monitoringServiceLogger.info({
          message: "Report card generated",
          meta: {
            reportCard,
            date: new Date(),
          },
        })
        return okAsync(reportCard)
      })
      .orElse(() => okAsync([]))
```